### PR TITLE
Optimise multiple Append and Prepend calls.

### PR DESCRIPTION
**Pull request by JonHanna**
Use a more compact form for multiple calls to `Append` and `Prepend`, that shares a collection of appended and/or prepended elements between instances.

Considering @stephentoub's comment at https://github.com/dotnet/corefx/pull/5947#discussion_r52251131

> I'm imaging at some point we might want to create some optimized paths for when Append, Prepend, and Concat are used repetitively.

This attempts to optimise multiple append and prepends, but not concatenation.

It would need some more tests at the very least before its ready for merging, but I'd appreciate any input on the approach taken prior to that.

### CODE REVIEW

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 1]
```diff
@@ -0,0 +1,715 @@
...
+        {
+            _store = new[] { first, second };
+            _count = 2;
+        }
+
+        public T this[int index] => _store[index >= 0 ? index : index + _store.Length];
+
+        public bool Append(T item, int number)
+        {
+            if (number != _count | number >= MaxLength)
```
Comment (svick) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Is using `|` instead of `||` here intentional micro-optimization? If that's the case, it seems to me it's a very unusual pattern in corefx, so maybe it should be explicitly pointed out by a comment?

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 2]
```diff
@@ -0,0 +1,715 @@
...
+                }
+
+                Debug.Assert(index == array.Length);
+                return array;
+            }
+
+            public override List<TSource> ToList()
+            {
+                int count = GetCount(onlyIfCheap: true);
+                List<TSource> list = count == -1 ? new List<TSource>() : new List<TSource>(count + _appCount + _preCount);
```
Comment (svick) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Would it make sense to use `new List<TSource>(_appCount + _preCount)` for the case when getting count isn't cheap? It should avoid some reallocations of the `List` array, at least as long as `_appCount + _preCount` is not smaller than `List<T>._defaultCapacity`.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs (hunk 1 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Its an as-a-rule rather than a measured micro-opt. As in "as a rule" just comparing integers will be cheaper than the cost of branching to avoid doing so, but you have to measure to know for sure.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs (hunk 2 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Good idea.

Comment (JonHanna) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: I set it to be `Math.Max(_appCount + _preCount, 4)` in this case, 4 being `List<T>._defaultCapacity`. Similarly for the single-item versions I set it to `new List<TSource(4)` since there's always going to be at least one `Add()` and hence on move from 0 to 4.

**Comment by JonHanna** [comment]
Test this please

**Comment by JonHanna** [comment]
Now why did dotnet-bot do that? (Was closing and re-opening seen as Spammy perhaps)

**Comment by JonHanna** [comment]
Test Innerloop Windows_NT Release Build and Test

(System.Net.Http.Functional.Tests.HttpClientHandlerTest.GetAsync_UnsupportedSSLVersion_Throws
Timeout caused cancelling? http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/windows_nt_release_prtest/4306/testReport/junit/System.Net.Http.Functional.Tests/HttpClientHandlerTest/GetAsync_UnsupportedSSLVersion_Throws_name___SSLv3___url___https___www_ssllabs_com_10300___/)

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 3]
```diff
@@ -0,0 +1,715 @@
...
+            public abstract AppendPrependIterator<TSource> Prepend(TSource element);
+
+            public abstract TSource[] ToArray();
+
+            public abstract List<TSource> ToList();
+
+            public abstract int GetCount(bool onlyIfCheap);
+        }
+
+        private sealed class Append1Iterator<TSource> : AppendPrependIterator<TSource>
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Rather than duplicating the type hierarchy for append vs prepend, my preference would be to carry a bool around and switch on it in the implementation.  It's not much more extra space required, the branch should be negligible compared to all of the interface calls being made, and the amount of code required and associated complexity should shrink non-trivially.

Comment (JonHanna) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: So have a single `AppendPrepend` with two collections?

Comment (stephentoub) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Ah, right, I hadn't gotten that far.  But, yeah, rather than having a single collection field, have two collection fields, one of which will likely be null.  We increase the size of the object by a single reference field, but I think cut back on complexity and duplication.  I think I'm suggesting just get rid of your AppendN and PrependN iterators, in favor of just using AppendPrependNIterator.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 4]
```diff
@@ -0,0 +1,715 @@
...
+
+                Debug.Assert(index == array.Length - 1);
+                array[index] = _element;
+                return array;
+            }
+
+            public override List<TSource> ToList()
+            {
+                int count = GetCount(onlyIfCheap: true);
+                List<TSource> list = count == -1 ? new List<TSource>(4) : new List<TSource>(count);
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Why are you explicitly specifying 4?  Let's just let List use its default heuristics, which will start with an empty array and grow to 4 the first time something is added.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs (hunk 3 above)
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: And also for this Append1 case, I'm suggesting having a single AppendPrepend1Iterator with an extra bool field rather than having two different Append1 and Prepend1 cases, and use that bool to determine whether the single _element is meant to be appended or prepended.

Comment (JonHanna) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: That would cut down the size, but unless it's all cut down to one class we're either left with still having the interface call, or adding to the number of type checks in the query method.

Of course the case with two collections can handle the Append1 and Prepend1 case too, and maybe the extra weight of that is worth the extra simplicity.

Comment (stephentoub) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: I'm not following.  Why would it change the number of type checks or interface calls?  It would add branches, yes, because you'd need to check whether the bool was false or true to determine whether to append or prepend the element... why is it more than that?

Comment (JonHanna) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Maybe it's me that's not following. I read you as suggesting an `AppendPrependNIterator` and an `AppendPrepend1Iterator` as separate classes, in which case to `Append` to one of those we need to either keep the abstract model in use here (so we don't lose the virtual call) or check for both of those types rather than just checking for one.

Comment (stephentoub) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: > I read you as suggesting an AppendPrependNIterator and an AppendPrepend1Iterator as separate classes

Yes.

> in which case to Append to one of those we need to either keep the abstract model in use here (so we don't lose the virtual call)

Yes.  I wasn't suggesting that my approach would be cheaper, but it also shouldn't add more virtual calls than what you already have, and it effectively cuts the amount of code in half.

Comment (JonHanna) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Ah. I see what I misread now.

>  …negligible compared to all of the interface calls being made

I misread that as suggesting that said calls would be compensatorily reduced.

**Comment by JonHanna** [comment]
Test Innerloop Windows_NT Debug Build and Test
Test Innerloop Windows_NT Release Build and Test
(Hung process, and SSL tests at http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/windows_nt_release_prtest/4672/ which might be a network fluke, will open an issue if it repeats).

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 5]
```diff
@@ -0,0 +1,472 @@
...
+                }
+
+                Debug.Assert(index == array.Length);
+                return array;
+            }
+
+            public override List<TSource> ToList()
+            {
+                int count = GetCount(onlyIfCheap: true);
+                List<TSource> list = count == -1 ? new List<TSource>(Math.Max(_appCount + _preCount, 4)) : new List<TSource>(count);
```
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: @stephentoub I dropped the `new List<TSource>(4)` from the other case, but I've left `Math.Max(_appCount + _preCount, 4)` as it seems that "the sum of the sizes of the two parts we know are adding in for sure, unless that's pretty low in which case a wee bit more" is a reasonable heuristic for the starting size, with 4 as a reasonable value for "wee bit more".

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 6]
```diff
@@ -0,0 +1,472 @@
...
+                _appending = appending;
+            }
+
+            public override Iterator<TSource> Clone() => new AppendPrepend1Iterator<TSource>(_source, _element, _appending);
+
+            public override bool MoveNext()
+            {
+                switch (_state)
+                {
+                    case 1:
```
Comment (hughbe) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: minor question: is it worth having these integers as private named constants for readability?

Comment (JonHanna) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: It's a linear progression through numbered states. I think the most readable labels are simply 1, 2, 3 etc.

Comment (hughbe) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: sure

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 7]
```diff
@@ -0,0 +1,472 @@
...
+                return false;
+            }
+
+            T[] store = _store;
+
+            if (store.Length == number)
+            {
+
+                T[] newStore = new T[number << 1];
+                for (int i = 0; i != store.Length; ++i)
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Does using `!=` as opposed to `<` here result in a performance improvement?

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 8]
```diff
@@ -0,0 +1,472 @@
...
+            }
+
+            store[number] = item;
+            ++_count;
+            return true;
+        }
+
+        public bool Prepend(T item, int number)
+        {
+            if (number != _count | number >= MaxLength)
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Shouldn't it be `> MaxLength`? I'm assuming since you named it such that 64 should be a valid value for `number`, right?

File: src/System.Linq/src/System/Linq/AppendPrepend.cs (hunk 1 above)
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Ditto here with my other line note; shouldn't it be `> MaxLength`?

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 9]
```diff
@@ -0,0 +1,472 @@
...
+                if (count == -1)
+                {
+                    return EnumerableHelpers.ToArray(this);
+                }
+
+                TSource[] array = new TSource[count];
+                int index = 0;
+                for (int i = -_preCount; i != 0; ++i, ++index)
+                {
+                    array[index] = _prepended[i];
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Perhaps `_prepended` should be captured into a local variable to save field accesses within the loop, like you did with `_store` below in `SharedCollection`.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 10]
```diff
@@ -0,0 +1,472 @@
...
+                foreach (TSource item in _source)
+                {
+                    array[index] = item;
+                    ++index;
+                }
+
+                Debug.Assert(index == array.Length - _appCount);
+                for (int i = 0; i != _appCount; ++i, ++index)
+                {
+                    array[index] = _appended[i];
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Ditto here for `_appended`.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 11]
```diff
@@ -0,0 +1,472 @@
...
+                return array;
+            }
+
+            public override List<TSource> ToList()
+            {
+                int count = GetCount(onlyIfCheap: true);
+                List<TSource> list = count == -1 ? new List<TSource>(Math.Max(_appCount + _preCount, 4)) : new List<TSource>(count);
+                for (int i = -_preCount; i != 0; ++i)
+                {
+                    list.Add(_prepended[i]);
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Ditto here; could be captured into a local variable outside the loop.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 12]
```diff
@@ -0,0 +1,472 @@
...
+                }
+
+                list.AddRange(_source);
+
+                for (int i = 0; i != _appCount; ++i)
+                {
+                    list.Add(_appended[i]);
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Ditto here; could be captured into a local variable outside the loop.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 13]
```diff
@@ -0,0 +1,472 @@
...
+            if (number != _count | number >= MaxLength)
+            {
+                return false;
+            }
+
+            T[] store = _store;
+
+            if (store.Length == number)
+            {
+                T[] newStore = new T[number << 1];
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Not that familiar with the CoreCLR internals/JIT, but maybe it already does this as an optimization? If not then maybe there should be a comment here explaining it's basically an optimized version of `* 2`.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 14]
```diff
@@ -0,0 +1,472 @@
...
+            {
+                return false;
+            }
+
+            T[] store = _store;
+
+            if (store.Length == number)
+            {
+
+                T[] newStore = new T[number << 1];
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Ditto here; leave a comment explaining it, or just put `* 2` if the JIT already does this as an optimization.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 15]
```diff
@@ -0,0 +1,472 @@
...
+            if (number != _count | number >= MaxLength)
+            {
+                return false;
+            }
+
+            T[] store = _store;
+
+            if (store.Length == number)
+            {
+
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Nit: unnecessary whitespace

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 16]
```diff
@@ -0,0 +1,472 @@
...
+            _count = 1;
+        }
+
+        public SharedCollection(T first, T second)
+        {
+            _store = new[] { first, second };
+            _count = 2;
+        }
+
+        public T this[int index] => _store[index >= 0 ? index : index + _store.Length];
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: From what I understand, this indexer allows you to write things like `collection[-1]` and have it return the last element of the collection. However, maybe it should just be simplified to `_store[index]`. The reason for this is it increases the likelihood of inlining by the JIT (also see [these](https://github.com/dotnet/corefx/blob/master/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray%601.cs#L126-L128) lines in `ImmutableArray`), and also having it like this introduces a new branch for every time we're using the indexer and not using this feature.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 17]
```diff
@@ -0,0 +1,472 @@
...
+                {
+                    list.Add(_appended[i]);
+                }
+
+                return list;
+            }
+        }
+    }
+
+    internal class SharedCollection<T>
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Maybe this could be made a `struct` to avoid the overheap of heap allocation? As far as I can see, it has two fields and is at most 12 bytes large.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 18]
```diff
@@ -0,0 +1,472 @@
...
+                }
+
+                return list;
+            }
+        }
+
+        private sealed class AppendPrependNIterator<TSource> : AppendPrependIterator<TSource>
+        {
+            private readonly SharedCollection<TSource> _prepended;
+            private readonly SharedCollection<TSource> _appended;
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: If you choose to make `SharedCollection` a struct, these fields need to be stripped of `readonly` since the values of `readonly` fields are always copied when accessing them even within the class. See [this](https://blogs.msdn.microsoft.com/ericlippert/2008/05/14/mutating-readonly-structs/) and [this](https://www.simple-talk.com/blogs/2010/05/19/why-enumerator-structs-are-a-really-bad-idea/) blog post for more.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs (hunk 7 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: I very much doubt it. Preference from a long time ago being more often using C++ iterator objects where `!=` might be defined for the type but `<` might not. (And longer ago again it might perhaps have made for a slight but measurable performance difference on old CPUs).

File: src/System.Linq/src/System/Linq/AppendPrepend.cs (hunk 17 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: But then it would be copied, and different copies would have different _counts. In order to have the shared behaviour it exists for that count would have to be wrapped in a reference type or boxed, and that defeats any gain in having it as a struct.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs (hunk 18 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: It's not so much that they're always copied when accessing them that would be the big problem, as they won't be copied back because they're readonly which would make it plain not work. Again though, we need a reference type here because we want to deliberately alias the same instance between different iterator instances.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs (hunk 17 above)
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: @JonHanna Ah, I hadn't seen that there were methods such as `Clone` or `Prepend` that passed the field into other methods. Never mind, then.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs (hunk 16 above)
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: In fact, maybe you should just ditch the indexer altogether and publicly expose the `_store` and `_count` fields. There's no need for enscapulation since this class is internal.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs (hunk 8 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Perhaps. It's rather arbitrary, in that on the one hand the larger we allow the collection to grow the better for the performance of the object produced on that call, but on the other hand the more memory wasted if we have the result of a call to `Prepend()` in long-lived memory, call chained `Prepend()` on that a lot and then let the result of those many calls be collected, leaving us with a massive `SharedCollection` where only a few elements are reachable by outside code. Being arbitrary one more or less doesn't make much difference either way. (Whether the number should perhaps be a considerable amount larger or smaller is another matter).

File: src/System.Linq/src/System/Linq/AppendPrepend.cs (hunk 17 above)
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: In that case, another nit: should be `sealed class`

File: src/System.Linq/src/System/Linq/AppendPrepend.cs (hunk 9 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Better still, perhaps `SharedCollection` should have a `CopyTo` to reduce indirection even further.

Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Aha, that's also a good idea. Then you could use `Array.Copy` which is way faster than an indexer.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs (hunk 13 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Another place where I'm just showing my age :wink: I expect it isn't terribly obscure though; there are at least a few places with the same thing in CoreFX and CoreCLR. (I think it doesn't normally do that optimisation, btw, and certainly there are differences between shifting and multiplication that won't come up here but do man it can only be considered equivalent in certain places. That said, while it may be my habit to bit-shift in cases like this, I don't think it's a habit worth acquiring).

File: src/System.Linq/src/System/Linq/AppendPrepend.cs (hunk 17 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: The sharing means that in the most expected case (a bunch of chained append and/or prepend calls so that each result of those only ever has append or prepend called on it once) there's a single collection used across all the results, which is also the only place where arrays are grown. It should indeed be sealed, as you say.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs (hunk 16 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: I like encapsulation because I trust myself only slightly more than other coders :wink:

I'll experiment with it, but I think I like the simplicity this provides. I may change my mind in the morning though.

Comment (JonHanna) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: With the fresh light of morning, no I don't think this is worth doing.
In the case of the lines you point out in `ImmutableArray` if that call was inlined into a method within which the jitter could tell statically that the index was definitely within range (an explicit range check had already been done) or would always be within range (e.g. it's moving from 0 to length - 1) then it can omit the range check, this needs the call to `Length` to have also be inlined into the calling method and the arithmetic to be relatively simple. None of the places where this is called match those criteria. The branching issue will be reduced by the change to the ToArray and ToList mentioned elsewhere. (The branching will still happen in MoveNext, but if it didn't happen there it would happen elsewhere, and at least the same branch will be picked in succession anyway, so the branch prediction will hopefully be correct).

File: src/System.Linq/src/System/Linq/AppendPrepend.cs (hunk 8 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: With the code fired up now, `>=` is the one that best matches this. `number` is the index (or negative index when prepending) so after the call the length can be as large as one more than `number`, so `>=` is what is wanted.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 19]
```diff
@@ -0,0 +1,496 @@
...
+        private T[] _store;
+        private int _count;
+
+        public SharedCollection(T item)
+        {
+            _store = new[] { item };
+            _count = 1;
+        }
+
+        public SharedCollection(T first, T second)
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Maybe it would be more readable if you had a `params T[] items` constructor here instead?

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 20]
```diff
@@ -0,0 +1,496 @@
...
+            }
+
+            T[] store = _store;
+
+            if (store.Length == number)
+            {
+                T[] newStore = new T[number << 1];
+                for (int i = 0; i != store.Length; ++i)
+                {
+                    newStore[i] = store[i];
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Maybe you could use `Array.Copy` here.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs (hunk 19 above)
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: > Maybe it would be more readable if you had a params T[] items constructor here instead?

That would allocate an array at the call site.

Comment (JonHanna) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: We could assign it straight to `_store` so it wouldn't add extra cost, but this makes it clear that it's only ever called with 1 or 2 elements.

File: src/System.Linq/src/System/Linq/AppendPrepend.cs [hunk 21]
```diff
@@ -0,0 +1,496 @@
...
+            }
+
+            T[] store = _store;
+
+            if (store.Length == number)
+            {
+                T[] newStore = new T[number << 1];
+                for (int i = 0; i != store.Length; ++i)
+                {
+                    newStore[number + i] = store[i];
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/AppendPrepend.cs]: Ditto here; maybe use `Array.Copy`? Not sure how fast it would be since the `MaxLength` of `store` appears to be 64, but for comparison `MemoryStream` uses `Buffer.BlockCopy` if the [length of the destionation > 8](https://github.com/dotnet/corefx/blob/master/src/System.IO/src/System/IO/MemoryStream.cs#L407), and does a for-loop otherwise.

**Comment by VSadov** [comment]
@dotnet-bot test Innerloop Windows_NT Debug Build and Test please
@dotnet-bot test Innerloop CentOS7.1 Release Build and Test and Test please

**Comment by VSadov** [comment]
I have concerns about attaching appended elements to iterator through which they are not reachable. It seems it can lead to unexpected retention.

```C#
// here a,b,c become GC-reachable from x, but not iterator-reachable
x.Append(a).Append(b).Append(c);  
```

I think prepend case could be handled by sharing a single-linked list. Then whatever is reachable via iteration is also what is GC-reachable.

Not sure if similar solution is possible for Append case.

**Comment by stephentoub** [comment]
> I have concerns about attaching appended elements to iterator through which they are not reachable

Excellent point.  I agree.

**Comment by JonHanna** [comment]
Yes. That's the reason for the limit on the size of the shared collection, but perhaps a linked-list would be better.

**Comment by stephentoub** [comment]
> That's the reason for the limit on the size of the shared collection

It's not necessarily the number that matters.  One of the appended items could be a giant instance.

**Comment by JonHanna** [comment]
That is true. A linked-list it shall be then.

**Comment by svick** [comment]
@JonHanna Are you saying you want to replace the current implementation, which is effectively a linked list of `IEnumerable`s, with an explicit linked list?

I think doing that would save virtual function calls when iterating the result, but at the cost of one more allocation per append. That doesn't sound like it would be worth it to me.

**Comment by JonHanna** [comment]
@svick there's a similar thing done with Concat. Maybe it will turn out to not be worth it upon trying, but I shall take a look.

**Comment by svick** [comment]
@JonHanna Sorry, I deleted my comment. I forgot that the current implementation means iterating is quadratic, so a linked list would probably still be worth it.

**Comment by JonHanna** [comment]
> Not sure if similar solution is possible for Append case.

I wasn't sure either when I first approached this, and so had rejected a linked-list approach, but if we keep track of the size of the list (which is worth doing anyway, for other optimisations) we can store the appended items in an array as they are being iterated, and so gather them up in a single O(n) sweep before an O(n) iteration through them, rather than having the quadratic behaviour.

File: src/System.Linq/src/System/Linq/Iterator.cs [hunk 22]
```diff
@@ -20,6 +20,8 @@ protected Iterator()
                 _threadId = Environment.CurrentManagedThreadId;
             }
 
+            protected bool RunningOnCreatingThread => _threadId == Environment.CurrentManagedThreadId;
```
Comment (VSadov) [inline src/System.Linq/src/System/Linq/Iterator.cs]: Is this used?

Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Iterator.cs]: I don't think it is after the last round of changes, so it can be reverted. I'm not having a lot of time for CoreFX right now, but I'll try to tidy this up ASAP.

**Comment by VSadov** [comment]
Looks like a good compromise. Append case needs an allocation of an array, but I do not see a lot of alternatives. Without that we have the n^2 problem when iterating.

RunningOnCreatingThread seems to be a method not used anywhere.

Otherwise LGTM.

**Comment by JonHanna** [comment]
That `RunningOnCreatingThread` cruft has been removed.
